splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/MergeJoinStrategy.java

```diff
@@ -218,33 +219,38 @@ private boolean mergeable(RowOrdering outerRowOrdering,
                                 IndexRowGenerator innerRowGenerator,
                                 OptimizablePredicateList predList,
                                 Optimizable innerTable) throws StandardException{
-        int[] keyColumnPositionMap = innerRowGenerator.baseColumnPositions();
+        boolean isIndexOnExpr = innerRowGenerator.isOnExpression();
+        int[] keyColumnPositionMap = isIndexOnExpr ? null : innerRowGenerator.baseColumnPositions();
         boolean[] keyAscending = innerRowGenerator.isAscending();
 
-        BitSet innerColumns = new BitSet(keyColumnPositionMap.length);
-        BitSet outerColumns = new BitSet(keyColumnPositionMap.length);
+        BitSet innerColumns = new BitSet(keyAscending.length);
+        BitSet outerColumns = new BitSet(keyAscending.length);
         for(int p = 0;p<predList.size();p++){
             Predicate pred = (Predicate)predList.getOptPredicate(p);
             if(pred.isJoinPredicate()) continue; //we'll deal with these later
             RelationalOperator relop=pred.getRelop();
             if(!(relop instanceof BinaryRelationalOperatorNode)) continue;
             if(relop.getOperator()==RelationalOperator.EQUALS_RELOP) {
-                int innerEquals = pred.hasEqualOnColumnList(keyColumnPositionMap, innerTable);
-                if (innerEquals >= 0) innerColumns.set(innerEquals);
-                else {
-
+                int innerEquals = isIndexOnExpr ? pred.hasEqualOnIndexExpression(innerTable)
+                                                : pred.hasEqualOnColumnList(keyColumnPositionMap, innerTable);
+                if (innerEquals >= 0) {
+                    innerColumns.set(innerEquals);
+                } else {
                     BinaryRelationalOperatorNode bron = (BinaryRelationalOperatorNode) relop;
-                    ValueNode vn = bron.getLeftOperand();
-                    if (!(vn instanceof ColumnReference))
-                        vn = bron.getRightOperand();
-                    if (!(vn instanceof ColumnReference)) continue;
-                    ColumnReference outerColumn = (ColumnReference) vn;
+                    int outerTableNum;
+                    int outerColNum;
+
                     /*
                      * We are still sortable if we have constant predicates on the first N keys on the outer
                      * side of the join, as long as we match the inner columns
                      */
-                    int outerTableNum = outerColumn.getTableNumber();
-                    int outerColNum = outerColumn.getColumnNumber();
+                    Pair<Integer, Integer> outerColumnInfo = getOuterColumnInfo(bron, innerTable, null);
+                    if (outerColumnInfo != null) {
+                        outerTableNum = outerColumnInfo.getFirst();
+                        outerColNum = outerColumnInfo.getSecond();
+                    } else {
+                        continue;
+                    }
```

Comment:
reorder
```
if (outerColumnInfo == null) continue;
int outerTableNum = outerColumnInfo.getFirst();
int outerColNum = outerColumnInfo.getSecond();
```
(now clear everything is initialized)

Reply:
Done.